### ISSUE TITLE
Use of specific class of tranformations

### ISSUE THREAD

**Issue by albanD** [qassemoquab/stnbhwd#1]
First, thank you very much for sharing this work.

I have been using your module to create a stn.
The main architecture I use is: (for example with a 10x10x3 input)
```lua
local localisation_network = nn.Sequential()
localisation_network:add(nn.Reshape(10*10*3))
local last_linear_module = nn.Linear(10*10*3, 6) -- (1)
localisation_network:add(last_linear_module)
localisation_network:add(nn.Reshape(2,3)) -- (2)
last_linear_module.weight:zero()
last_linear_module.bias = torch.Tensor{1,0,0,0,1,0}

local ct = nn.ConcatTable()
local branch1 = nn.Identity()
local branch2 = nn.Sequential()
branch2:add(localisation_network)
-- (3)
branch2:add(nn.AffineGridGeneratorBHWD(10,10))
ct:add(branch1)
ct:add(branch2)

local st_module = nn.Sequential()
st_module:add(ct)
st_module:add(nn.BilinearSamplerBHWD())
```
Though I would like to have the ability to use specific class of transformations (just a scale or just a translation...).

To achieve this, I modify the `last_linear_module` (1) to output the proper number of parameter (1 for scale, 2 for translation...), remove the final reshape of the `localisation_network` (2) and add at (3) a new module `AffineParamsGenerator`

This module takes as input a mode and and the corresponding number of parameters and outputs the (B,2,3) matrix corresponding. For example given `translation` and 2 inputs (tx, ty), outputs the proper parameter {{1,0,tx},{0,1,ty}} matrix for the `AffineGridGeneratorBHWD`.

##### Question

* Does this pipeline sounds correct to you? (if yes, do you want me to add more informations in the readme on how to create a stn from the modules given in this repo?)
* Did you work on this kind of module like `AffineParamsGenerator` to use only specific transformation? Were you using the same structure or do you think of something simpler?
* If yes, can you give me an idea of how you compute the gradient for this module in the case for example of isotropic scale where the transformation matrix is {{s,0,0},{0,s,0}}. Can we just sum the gradients returned by the AffineGridGenerator at the corresponding positions?

**Comment by qassemoquab** [comment]
The pipeline sounds good. I plan to do a demo myself, but help is always appreciated of course.

I didn't put constraints on the transform matrices, but I imagine you can simply zero out the parts of the transform matrices and the gradients that correspond to your constraints. I would do it in-place in a separate module just before AffineGridGenerator.

AffineGridGenerator already does this sum over x,y positions so you shouldn't worry about it, but I might have misunderstood what you meant.

**Comment by albanD** [comment]
For example:
* In the case of translation:
The localisation_network return two parameters `tx` and `ty`.
The corresponfing transform matrix will be {{1,0,tx},{0,1,ty}}
The gradient for `tx` will be `grid_generator_module.gradInput[batch][1][3]` and for `ty` `grid_generator_module.gradInput[batch][2][3]`. This case is ok.
* In the case of isotropic scaling:
The localisation_network returns one parameter `s`
The corresponding transform matrix will be {{s,0,0},{0,s,0}}
As you can see the gradient on s does not comme directly in the same way as for the translation since it is used at two positions in the transform matrix.
What should be the gradient for s then? `grid_generator_module.gradInput[batch][1][1] + grid_generator_module.gradInput[batch][2][2]` ?
* To generalize this idea, what should the gradient be in the case where the localisation_network output an angle `theta` for a rotation?

**Comment by qassemoquab** [comment]
I see what you mean !

For isotropic scaling :
- localisation network should output a single value
- one module should make it a [s 0 0 0 s 0] matrix
- that module, at backprop time, should sum the gradients and pass it back to the localisation net

For theta-based rotation :
- localisation network should output a single value
- one module should make it a [cos t, sin t, 0, -sin t, cos t, 0] matrix (note sure about the signs of sin)
- that module, at backprop time, should element-wise-multiply by the corresponding gradient of cos/sin at theta, sum the gradients and pass it back to the localisation net

Or, to make it as a sequence (each "=>" would be a module)
=> theta (differentiable wrt to theta : it is the output of the localization network)
=> [theta, theta, 0, theta, theta, 0] (differentiable wrt. theta : you just have to sum the gradients)
=> [cos theta, sin theta, 0, -sin theta, cos theta, 0]  (differentiable wrt. theta : you just have to perform - an element-wise cos' / sin' on the input matrix of that module
- and element-wise multiply by gradOutput to get gradInput
- zero out the last column, that corresponds to the translation parameters )
=> AffineGridGenerator (differentiable)

Does that make sense ?

**Comment by qassemoquab** [comment]
eg. for Isotropic Scaling you could plug this module just before the grid generator :

```
local ISGG, parent = torch.class('nn.IsotropicScaling', 'nn.Module')

function ISGG:__init()
   parent.__init(self)
end

local function addOuterDim(t)
   local sizes = t:size()
   local newsizes = torch.LongStorage(sizes:size()+1)
   newsizes[1]=1
   for i=1,sizes:size() do
      newsizes[i+1]=sizes[i]
   end
   return t:view(newsizes)
end

function ISGG:updateOutput(_input)
   local input
   if _input:nDimension()==1 then
      input = addOuterDim(_input)
   else
      input = _input
   end

   self.output:resize(input:size(1), 2, 3)
   self.output:select(3, 2):select(2,2):copy(input)
   self.output:select(3, 1):select(2,1):copy(input)

   if _input:nDimension()==1 then
      self.output = self.output:select(1,1)
   end

   return self.output
end

function ISGG:updateGradInput(_input, _gradOutput)
   local input, gradOutput
   if _input:nDimension()==1 then
      input = addOuterDim(_input)
      gradOutput = addOuterDim(_gradOutput)
   else
      input = _input
      gradOutput = _gradOutput
   end

   self.gradInput:resizeAs(input):zero()
   self.gradInput:add(_gradOutput:select(3,1):select(2,1))
   self.gradInput:add(_gradOutput:select(3,2):select(2,2))

   if _input:nDimension()==1 then
      self.gradInput = self.gradInput:select(1,1)
   end
end
```

**Comment by albanD** [comment]
Ok, exactly what I had in mind.

Do you think this kind of module would be relevant to add to this repo (a more general version), or it should be done by the user depending of its needs?
The problem being that a more general version will just be an enumeration of special cases.

PS: you have a typo in the updateGradInput:
```self.gradInput:add(_gradOutput:select(3,1):select(2,1))``` shoud be ```self.gradInput:add(gradOutput:select(3,1):select(2,1))```

**Comment by qassemoquab** [comment]
Thanks, I fixed that. 

That thing is recent, so I don't know what all the use-cases are. But if you come up with some constraint module that is relevant to your task, we can add it there of course.

**Comment by albanD** [comment]
The new nn.AffineTransformMatrixGenerator module allows the restriction of transformation.
Closing.
